### DISCUSSION TITLE
MOU with risk as an objective and opt_risk ?

### DISCUSSION THREAD

**Discussion by apryet** [usgs/pestpp#286] [Q&A] [answered]
Hello all,

After previous runs where risk was specified, I am now running PESTPP-MOU with risk as an objective, so I removed the opt_risk option. 
This is what  PESTPP-MOU yields: 
```
...
  ---  initializing MOEA process  ---  
...using 'nsga2' env selector
...using binary tournament mating pool selector
...saving generation specific populations and archives every nth generation 1
'opt_dec_var_groups' passed, using 65 adjustable parameters as decision variables
...number of decision variables:  65
...max run fail:  1
...'opt_chance_points' = SINGLE, evaluting chance at representative point
...using the following observations as objectives: 
                   DEFICIT_TOT   minimize
                      TOT_PUMP   maximize
...using the following prior info eqs as objectives: 
                        _RISK_   maximize
...'mou_risk_objective' is true, using _RISK_ decision variable as risk in chance calcs
...39422 obs-based constraints, see rec file for listing
...1 pi-based constraints, see rec file for listing
...opt_risk = 0.95, using stack-based chance constraints with 34 realizations
...
```
See the last line, I guess 0.95 is the default value for opt_risk, but it should have no effect when running with risk as an objective...right ?

My comparison of specified risk with opt_risk=0.66 does not really align with the corresponding value when running with risk as an objective : 

![pareto_risk_as_objective](https://github.com/usgs/pestpp/assets/4189108/17b2df83-bfa8-4240-b1fb-c0a640466077)

I understand that PESTPP-MOU, when running with the "risk as an objective" option  considers the adjustable _risk_ value to identify the dominated solutions in the space of objectives. But what about the constraints ? Does it consider the default value of opt_risk (0.95) to offset simulated constraint values and identify solutions as feasible ?

This is not what I expect, but that may be (among others...) one explanation of the mismatch between specified risk and risk as an objective solutions. 

Thank you for your help !

Alexandre

**Comment by jtwhite79** [comment, marked as answer]
The way that risk-as-an-objective works is that under the hood we need a risk value != 0.5 (this disables the chance processes). So if mou see that you want to use risk as an objective, it [resets risk to 0.95](https://github.com/usgs/pestpp/blob/7d094ae7f82ca74922b10fa2d37d67d620de9df9/src/libs/pestpp_common/MOEA.cpp#L2037) only to trigger the chance processes.  That actual value of risk is not used, instead mou is seeking to maximize risk along with seek the extrema of the other objectives.  As far as constraints, if they are "model based" (ie observations), then they are also subject to chances. But PI-based constraints/objectives are deterministic. And those model-based constraints receive the same risk-as-an-obj treatment and the model-based objectives.

The main reason I can think of for that for that discrepancy between the fixed risk and risk obj is that if you use risk as an obj, that great increases the complexity of the optimization problem, meaning you'll probably need a lot more generations - did you run more generations with risk as an obj?   The other issue is that when you add an objective, the dimensionality and orientation of any "null space" or non-uniqueness in the relation between decision variables and objectives/constraints will change, meaning that you can also get a different pareto surface (esp in nonlinear real-world settings).  Anyway, just pontificating now...

**Comment by apryet** [comment]
All right, thank you @jtwhite79  ! If objectives and constraints are subject to the same treatment (which is a good thing), the discrepancy could indeed be related to convergence or equifinality. The figure I provided was with the same (small) number of iterations : 10, but MOU is still running, I'll see if patience addresses the problem !

**Comment by apryet** [comment]
Indeed @jtwhite79 , looks better 2 days later, with more iterations (here 25). 

![Figure_1](https://github.com/usgs/pestpp/assets/4189108/393a7d3f-786c-4910-a8eb-5689f1b1cfed)

**Reply by jtwhite79** [reply]
ah yeah!
